File: skills/gemini-tts/scripts/tts.py
#!/usr/bin/env python3
"""Generate speech from text using Gemini TTS API.

Usage:
    python tts.py "Hello, world!"
    python tts.py "Welcome!" --voice Puck --output welcome.wav
    python tts.py "Conversation text" --multi --speakers "Joe:Kore,Jane:Puck"
    python tts.py "Long text" --stream

Requirements:
    pip install google-genai
"""

import argparse
import os
import sys
import wave

# Load .env file if present
try:
    from dotenv import load_dotenv

    load_dotenv()
except ImportError:
    pass


def get_client():
    """Get Gemini API client."""
    try:
        from google import genai
    except ImportError:
        print("Error: google-genai not installed. Run: pip install google-genai")
        sys.exit(1)

    api_key = os.environ.get("GOOGLE_API_KEY") or os.environ.get("GEMINI_API_KEY")
    if not api_key:
        print("Error: Set GOOGLE_API_KEY or GEMINI_API_KEY environment variable")
        sys.exit(1)

    return genai.Client(api_key=api_key)


def save_wav(filename: str, audio_data: bytes, rate: int = 24000):
    """Save raw PCM audio data to WAV file."""
    with wave.open(filename, "wb") as wf:
        wf.setnchannels(1)
        wf.setsampwidth(2)  # 16-bit
        wf.setframerate(rate)
        wf.writeframes(audio_data)


def generate_tts(
    text: str,
    voice: str = "Kore",
    output: str = "output.wav",
    model: str = "gemini-2.5-flash-preview-tts",
    stream: bool = False,
    speakers: dict | None = None,
) -> str:
    """Generate speech from text.

    Args:
        text: Text to convert to speech
        voice: Voice name (for single speaker)
        output: Output WAV file path
        model: TTS model ID
        stream: Use streaming for long text
        speakers: Dict mapping speaker names to voices (for multi-speaker)

    Returns:
        Path to saved audio file
    """
    from google.genai import types

    client = get_client()

    # Build speech config
    if speakers:
        speaker_configs = [
            types.SpeakerVoiceConfig(
                speaker=name,
                voice_config=types.VoiceConfig(
                    prebuilt_voice_config=types.PrebuiltVoiceConfig(voice_name=v)
                ),
            )
            for name, v in speakers.items()
        ]
        speech_config = types.SpeechConfig(
            multi_speaker_voice_config=types.MultiSpeakerVoiceConfig(
                speaker_voice_configs=speaker_configs
            )
        )
    else:
        speech_config = types.SpeechConfig(
            voice_config=types.VoiceConfig(
                prebuilt_voice_config=types.PrebuiltVoiceConfig(voice_name=voice)
            )
        )

    config = types.GenerateContentConfig(
        response_modalities=["AUDIO"],
        speech_config=speech_config,
    )

    contents = [types.Content(role="user", parts=[types.Part.from_text(text=text)])]

    all_audio = b""

    if stream:
        print("Streaming audio...")
        for chunk in client.models.generate_content_stream(
            model=model, contents=contents, config=config
        ):
            if chunk.candidates and chunk.candidates[0].content.parts:
                part = chunk.candidates[0].content.parts[0]
                if hasattr(part, "inline_data") and part.inline_data:
                    all_audio += part.inline_data.data
    else:
        response = client.models.generate_content(
            model=model, contents=contents, config=config
        )
        if response.candidates and response.candidates[0].content.parts:
            part = response.candidates[0].content.parts[0]
            if hasattr(part, "inline_data") and part.inline_data:
                all_audio = part.inline_data.data

    if all_audio:
        if not output.endswith(".wav"):
            output += ".wav"
        save_wav(output, all_audio)
        print(f"Saved: {output}")
        return output
    else:
        print("Error: No audio generated")
        sys.exit(1)


def parse_speakers(speakers_str: str) -> dict:
    """Parse speaker string like 'Joe:Kore,Jane:Puck' into dict."""
    speakers = {}
    for pair in speakers_str.split(","):
        if ":" in pair:
            name, voice = pair.split(":", 1)
            speakers[name.strip()] = voice.strip()
    return speakers


def main():
    parser = argparse.ArgumentParser(
        description="Generate speech from text using Gemini TTS"
    )
    parser.add_argument("text", help="Text to convert to speech")
    parser.add_argument(
        "--voice",
        "-v",
        default="Kore",
        help="Voice name (default: Kore). Options: Kore, Puck, Charon, Fenrir, Aoede, Zephyr, Sulafat",
    )
    parser.add_argument(
        "--output",
        "-o",
        default="output.wav",
        help="Output WAV file path (default: output.wav)",
    )
    parser.add_argument(
        "--model",
        "-m",
        default="gemini-2.5-flash-preview-tts",
        help="Model ID (default: gemini-2.5-flash-preview-tts)",
    )
    parser.add_argument(
        "--stream", "-s", action="store_true", help="Use streaming for long text"
    )
    parser.add_argument(
        "--speakers", help="Multi-speaker mapping: 'Speaker1:Voice1,Speaker2:Voice2'"
    )

    args = parser.parse_args()

    speakers = parse_speakers(args.speakers) if args.speakers else None

    try:
        generate_tts(
            text=args.text,
            voice=args.voice,
            output=args.output,
            model=args.model,
            stream=args.stream,
            speakers=speakers,
        )
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: skills/gemini-tts/scripts/tts.js
#!/usr/bin/env node
"use strict";

import { mkdir, writeFile } from "node:fs/promises";
import path from "node:path";
import process from "node:process";
import { parseArgs } from "node:util";

import dotenv from "dotenv";
import { GoogleGenAI } from "@google/genai";

dotenv.config();

const DEFAULT_VOICE = "Kore";
const DEFAULT_OUTPUT_DIR = "audio/";
const DEFAULT_OUTPUT_NAME = "tts_output";
const DEFAULT_MODEL = "gemini-2.5-flash-preview-tts";

function printUsage() {
  console.log(`Generate speech from text using Gemini TTS API.

Usage:
  node scripts/tts.js "Hello, world!"
  node scripts/tts.js "Welcome!" --voice Puck --output welcome
  node scripts/tts.js "Conversation text" --speakers "Joe:Kore,Jane:Puck"
  node scripts/tts.js "Long text" --stream
  node scripts/tts.js "Custom folder" --output-dir ./my-audio/

Options:
  --voice, -v        Voice name (default: ${DEFAULT_VOICE})
  --output-dir       Output directory (default: ${DEFAULT_OUTPUT_DIR})
  --output, -o       Base output name (default: ${DEFAULT_OUTPUT_NAME})
  --no-timestamp     Disable timestamp in filename
  --model, -m        TTS model ID (default: ${DEFAULT_MODEL})
  --stream, -s       Use streaming mode
  --speakers         Multi-speaker map: "Joe:Kore,Jane:Puck"
  --help, -h         Show this help message

Requirements:
  npm install @google/genai@latest dotenv@latest
`);
}

function getApiKey() {
  const apiKey = process.env.GOOGLE_API_KEY || process.env.GEMINI_API_KEY;
  if (!apiKey) {
    throw new Error("Set GOOGLE_API_KEY or GEMINI_API_KEY environment variable");
  }
  return apiKey;
}

function parseSpeakers(input) {
  const entries = [];
  for (const pair of input.split(",")) {
    if (!pair.includes(":")) continue;
    const [speaker, voice] = pair.split(":", 2).map((v) => v.trim());
    if (speaker && voice) {
      entries.push({ speaker, voice });
    }
  }
  return entries;
}

function decodeAudioData(data) {
  if (!data) return null;
  if (typeof data === "string") return Buffer.from(data, "base64");
  if (data instanceof Uint8Array) return Buffer.from(data);
  if (Array.isArray(data)) return Buffer.from(data);
  return null;
}

function buildWavBuffer(pcmData, sampleRate = 24000) {
  const channels = 1;
  const bitsPerSample = 16;
  const byteRate = sampleRate * channels * (bitsPerSample / 8);
  const blockAlign = channels * (bitsPerSample / 8);
  const dataSize = pcmData.length;
  const buffer = Buffer.alloc(44 + dataSize);

  buffer.write("RIFF", 0);
  buffer.writeUInt32LE(36 + dataSize, 4);
  buffer.write("WAVE", 8);
  buffer.write("fmt ", 12);
  buffer.writeUInt32LE(16, 16);
  buffer.writeUInt16LE(1, 20);
  buffer.writeUInt16LE(channels, 22);
  buffer.writeUInt32LE(sampleRate, 24);
  buffer.writeUInt32LE(byteRate, 28);
  buffer.writeUInt16LE(blockAlign, 32);
  buffer.writeUInt16LE(bitsPerSample, 34);
  buffer.write("data", 36);
  buffer.writeUInt32LE(dataSize, 40);
  pcmData.copy(buffer, 44);
  return buffer;
}

function buildFilename(baseName, useTimestamp) {
  if (!useTimestamp) {
    return baseName.endsWith(".wav") ? baseName : `${baseName}.wav`;
  }
  const now = new Date();
  const yyyy = now.getFullYear();
  const mm = String(now.getMonth() + 1).padStart(2, "0");
  const dd = String(now.getDate()).padStart(2, "0");
  const hh = String(now.getHours()).padStart(2, "0");
  const mi = String(now.getMinutes()).padStart(2, "0");
  const ss = String(now.getSeconds()).padStart(2, "0");
  return `${baseName}_${yyyy}${mm}${dd}_${hh}${mi}${ss}.wav`;
}

async function generateTts({ text, voice, outputDir, outputName, model, stream, speakers, useTimestamp }) {
  const ai = new GoogleGenAI({ apiKey: getApiKey() });

  let speechConfig;
  if (speakers.length > 0) {
    speechConfig = {
      multiSpeakerVoiceConfig: {
        speakerVoiceConfigs: speakers.map((entry) => ({
          speaker: entry.speaker,
          voiceConfig: {
            prebuiltVoiceConfig: { voiceName: entry.voice },
          },
        })),
      },
    };
  } else {
    speechConfig = {
      voiceConfig: {
        prebuiltVoiceConfig: { voiceName: voice },
      },
    };
  }

  const request = {
    model,
    contents: [{ role: "user", parts: [{ text }] }],
    config: {
      responseModalities: ["AUDIO"],
      speechConfig,
    },
  };

  const chunks = [];
  if (stream) {
    console.log("Streaming audio...");
    const responseStream = await ai.models.generateContentStream(request);
    for await (const chunk of responseStream) {
      const part = chunk.candidates?.[0]?.content?.parts?.[0];
      const audioChunk = decodeAudioData(part?.inlineData?.data);
      if (audioChunk) {
        chunks.push(audioChunk);
      }
    }
  } else {
    const response = await ai.models.generateContent(request);
    const part = response.candidates?.[0]?.content?.parts?.[0];
    const audioData = decodeAudioData(part?.inlineData?.data);
    if (audioData) {
      chunks.push(audioData);
    }
  }

  if (chunks.length === 0) {
    throw new Error("No audio generated");
  }

  await mkdir(outputDir, { recursive: true });
  const filename = buildFilename(outputName, useTimestamp);
  const fullPath = path.join(outputDir, filename);
  const pcmData = Buffer.concat(chunks);
  const wavData = buildWavBuffer(pcmData, 24000);
  await writeFile(fullPath, wavData);
  console.log(`Saved: ${fullPath}`);
  return fullPath;
}

function parseCli() {
  const { values, positionals } = parseArgs({
    options: {
      voice: { type: "string", short: "v", default: DEFAULT_VOICE },
      "output-dir": { type: "string", default: DEFAULT_OUTPUT_DIR },
      output: { type: "string", short: "o", default: DEFAULT_OUTPUT_NAME },
      "no-timestamp": { type: "boolean", default: false },
      model: { type: "string", short: "m", default: DEFAULT_MODEL },
      stream: { type: "boolean", short: "s", default: false },
      speakers: { type: "string" },
      help: { type: "boolean", short: "h", default: false },
    },
    allowPositionals: true,
  });

  if (values.help) {
    printUsage();
    process.exit(0);
  }

  if (positionals.length === 0) {
    printUsage();
    process.exit(1);
  }

  return {
    text: positionals.join(" "),
    voice: values.voice,
    outputDir: values["output-dir"],
    outputName: values.output,
    model: values.model,
    stream: values.stream,
    speakers: values.speakers ? parseSpeakers(values.speakers) : [],
    useTimestamp: !values["no-timestamp"],
  };
}

async function main() {
  try {
    const options = parseCli();
    await generateTts(options);
  } catch (error) {
    console.error(`Error: ${error instanceof Error ? error.message : String(error)}`);
    process.exit(1);
  }
}

await main();
